src/mbgl/renderer/style_diff.cpp

```diff
@@ -43,14 +38,27 @@ StyleDifference<T> diff(const std::vector<T>& a, const std::vector<T>& b) {
     return result;
 }
 
+ImageDifference diffImages(const std::vector<ImmutableImage>& a,
+                           const std::vector<ImmutableImage>& b) {
+    return diff(a, b, [] (const ImmutableImage& lhs, const ImmutableImage& rhs) {
+        return lhs->id == rhs->id;
+    });
+}
+
 SourceDifference diffSources(const std::vector<ImmutableSource>& a,
                              const std::vector<ImmutableSource>& b) {
-    return diff(a, b);
+    return diff(a, b, [] (const ImmutableSource& lhs, const ImmutableSource& rhs) {
+        return std::tie(lhs->id, lhs->type)
+            == std::tie(rhs->id, rhs->type);
```

Comment:
Similarly, this breaks down when id/type match, but different styles use the same source name (`composite` is the leading contender) to point to different underlying tilesets.